atlasdb-config/src/main/java/com/palantir/atlasdb/factory/timelock/TimestampCorroboratingTimelockService.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * (c) Copyright 2018 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.factory.timelock;
+
+import java.util.function.Supplier;
+import java.util.function.ToLongFunction;
+
+import com.palantir.lock.v2.AutoDelegate_TimelockService;
+import com.palantir.lock.v2.IdentifiedTimeLockRequest;
+import com.palantir.lock.v2.StartAtlasDbTransactionResponse;
+import com.palantir.lock.v2.TimelockService;
+import com.palantir.timestamp.TimestampRange;
+
+/**
+ * A timelock service decorator for introducing runtime validity checks on received timestamps.
+ */
+public class TimestampCorroboratingTimelockService implements AutoDelegate_TimelockService {
+    static final String CLOCKS_WENT_BACKWARDS_MESSAGE =
+            "Expected timestamp to be greater than %s, but a fresh timestamp was %s!";
+
+    private final TimelockService delegate;
+    private volatile long lowerBound = Long.MIN_VALUE;
+
+    private TimestampCorroboratingTimelockService(TimelockService delegate) {
+        this.delegate = delegate;
+    }
+
+    public static TimelockService create(TimelockService delegate) {
+        return new TimestampCorroboratingTimelockService(delegate);
+    }
+
+    @Override
+    public TimelockService delegate() {
+        return delegate;
+    }
+
+    @Override
+    public long getFreshTimestamp() {
+        return checkAndUpdate(delegate::getFreshTimestamp, x -> x, x -> x);
```

Comment:
not really sure we need the functional approach, it (imo) unnecessarily complicates `checkAndUpdate`. It should be as simple as:
```java
public long getFreshTimestamp() {
  long delegateTimestamp = delegate.getFreshTimestamp();
  checkAndUpdate(delegateTimestamp, delegateTimestamp);
  return delegateTimestamp;
}

public TimestampRange getFreshTimestamps(numRequested) {
  TimestampRange range = delegate.getFreshTimestamps(numRequested);
  checkAndUpdate(range.getLowerBound(), range.getUpperBound());
  return range;
}

public StartAtlasDbTransactionResponse startAtlasDbTransaction(request) {
  StartAtlaqsDbTransactionResponse response = delegate.startAtlasDbTransaction(request);
  long timestamp = response.freshTimestamp();
  checkAndUpdate(timestamp, timestamp);
  return response;
}
```

Reply:
That implementation is very racey, Gokcan's is not!

Reply:
Gokcan's takes a snapshot of the lower bound before it calls the server. Yours will fail because two concurrent requests will race with each other (A updates the lower bound, then B does).